driver-core/src/test/unit/com/mongodb/internal/connection/ConnectionPoolAsyncTest.java

```diff
@@ -0,0 +1,394 @@
+        }
+        Object next = eventsIterator.next();
+        assertEquals(expectedType, next.getClass());
+        return expectedType.cast(next);
+    }
+
+    private Callable<Exception> createCallable(final BsonDocument operation) {
+        String name = operation.getString("name").getValue();
+        if (name.equals("checkOut")) {
+            FutureResultCallback<InternalConnection> callback = new FutureResultCallback<InternalConnection>();
+            return new Callable<Exception>() {
+                @Override
+                public Exception call() {
+                    try {
+                        pool.getAsync(new SingleResultCallback<InternalConnection>() {
+                            @Override
+                            public void onResult(final InternalConnection connection, final Throwable t) {
+                                if (t != null) {
+                                    callback.onResult(null, t);
+                                } else {
+                                    if (operation.containsKey("label")) {
+                                        connectionMap.put(operation.getString("label").getValue(), connection);
+                                    }
+                                    callback.onResult(connection, null);
+                                }
+                            }
+                        });
+                        callback.get();
+                        return null;
+                    } catch (Exception e) {
+                        return e;
+                    }
+                }
+            };
+        } else if (name.equals("checkIn")) {
+            return new Callable<Exception>() {
+                @Override
+                public Exception call() {
+                    try {
+                        InternalConnection connection = connectionMap.get(operation.getString("connection").getValue());
+                        connection.close();
+                        return null;
+                    } catch (Exception e) {
+                        return e;
+                    }
+                }
+            };
+        } else {
+            throw new UnsupportedOperationException("Operation " + name + " not supported");
+        }
+    }
+
+    @Parameterized.Parameters(name = "{0}: {1}")
+    public static Collection<Object[]> data() throws URISyntaxException, IOException {
+        List<Object[]> data = new ArrayList<Object[]>();
+        for (File file : JsonPoweredTestHelper.getTestFiles("/connection-monitoring-and-pooling")) {
+            if (!ignoreTestFile(file)) {
+                BsonDocument testDocument = JsonPoweredTestHelper.getTestDocument(file);
+                data.add(new Object[]{file.getName(), testDocument.getString("description").getValue(), testDocument});
+            }
+        }
+        return data;
+    }
+
+    private static boolean ignoreTestFile(final File file) {
```

Comment:
These tests always fail in the async test runner, either because a timeout error is emitted or a Future is not executed in the executor service when `getAsync()` is called.

Reply:
Are you working on debugging this?

Reply:
I have figured out the issues with the failing tests in the async test runner. The ConnectionCheckOutFailed events were emitted too soon. I will push a new commit soon that will resolve these test issues.